File: idea/tests/org/jetbrains/kotlin/idea/codeInsight/unwrap/AbstractUnwrapRemoveTest.java
/*
 * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
 * that can be found in the license/LICENSE.txt file.
 */

package org.jetbrains.kotlin.idea.codeInsight.unwrap;

import com.intellij.codeInsight.unwrap.Unwrapper;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.util.Condition;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.psi.PsiElement;
import com.intellij.testFramework.LightCodeInsightTestCase;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.kotlin.test.InTextDirectivesUtils;

import java.io.File;
import java.util.List;

public abstract class AbstractUnwrapRemoveTest extends LightCodeInsightTestCase {
    public void doTestExpressionRemover(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinExpressionRemover.class);
    }

    public void doTestThenUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinThenUnwrapper.class);
    }

    public void doTestElseUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinElseUnwrapper.class);
    }

    public void doTestElseRemover(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinElseRemover.class);
    }

    public void doTestLoopUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinLoopUnwrapper.class);
    }

    public void doTestTryUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinTryUnwrapper.class);
    }

    public void doTestCatchUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinCatchUnwrapper.class);
    }

    public void doTestCatchRemover(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinCatchRemover.class);
    }

    public void doTestFinallyUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinFinallyUnwrapper.class);
    }

    public void doTestFinallyRemover(@NotNull String path) throws Exception {
        doTest(path, KotlinUnwrappers.KotlinFinallyRemover.class);
    }

    public void doTestLambdaUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinLambdaUnwrapper.class);
    }

    public void doTestFunctionParameterUnwrapper(@NotNull String path) throws Exception {
        doTest(path, KotlinFunctionParameterUnwrapper.class);
    }

    private void doTest(@NotNull String path, final Class<? extends Unwrapper> unwrapperClass) throws Exception {
        configureByFile(path);

        String fileText = FileUtil.loadFile(new File(path), true);

        String isApplicableString = InTextDirectivesUtils.findStringWithPrefixes(fileText, "// IS_APPLICABLE: ");
        boolean isApplicableExpected = isApplicableString == null || isApplicableString.equals("true");

        String option = InTextDirectivesUtils.findStringWithPrefixes(fileText, "// OPTION: ");
        Integer optionIndex = option != null ? Integer.parseInt(option) : 0;

        List<Pair<PsiElement, Unwrapper>> unwrappersWithPsi =
                new KotlinUnwrapDescriptor().collectUnwrappers(getProject(), getEditor(), getFile());

        if (isApplicableExpected) {
            final Pair<PsiElement, Unwrapper> selectedUnwrapperWithPsi = unwrappersWithPsi.get(optionIndex);
            assertEquals(unwrapperClass, selectedUnwrapperWithPsi.second.getClass());

            final PsiElement first = selectedUnwrapperWithPsi.first;

            ApplicationManager.getApplication().runWriteAction(new Runnable() {
                @Override
                public void run() {
                    selectedUnwrapperWithPsi.second.unwrap(getEditor(), first);
                }
            });

            checkResultByFile(path + ".after");
        } else {
            assertTrue(
                    ContainerUtil.and(unwrappersWithPsi, new Condition<Pair<PsiElement, Unwrapper>>() {
                        @Override
                        public boolean value(Pair<PsiElement, Unwrapper> pair) {
                            return pair.second.getClass() != unwrapperClass;
                        }
                    })
            );
        }
    }

    @NotNull
    @Override
    protected String getTestDataPath() {
        return "";
    }
}

File: kotlin-ide/idea/test/org/jetbrains/kotlin/idea/codeInsight/unwrap/AbstractUnwrapRemoveTest.kt
/*
 * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
 * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
 */
package org.jetbrains.kotlin.idea.codeInsight.unwrap

import com.intellij.codeInsight.unwrap.Unwrapper
import com.intellij.openapi.application.runWriteAction
import com.intellij.openapi.command.executeCommand
import org.jetbrains.kotlin.idea.test.KotlinLightCodeInsightFixtureTestCase
import org.jetbrains.kotlin.test.InTextDirectivesUtils
import java.io.File

abstract class AbstractUnwrapRemoveTest : KotlinLightCodeInsightFixtureTestCase() {
    fun doTestExpressionRemover(unused: String) = doTest(KotlinUnwrappers.KotlinExpressionRemover::class.java)
    fun doTestThenUnwrapper(unused: String) = doTest(KotlinUnwrappers.KotlinThenUnwrapper::class.java)
    fun doTestElseUnwrapper(unused: String) = doTest(KotlinUnwrappers.KotlinElseUnwrapper::class.java)
    fun doTestElseRemover(unused: String) = doTest(KotlinUnwrappers.KotlinElseRemover::class.java)
    fun doTestLoopUnwrapper(unused: String) = doTest(KotlinUnwrappers.KotlinLoopUnwrapper::class.java)
    fun doTestTryUnwrapper(unused: String) = doTest(KotlinUnwrappers.KotlinTryUnwrapper::class.java)
    fun doTestCatchUnwrapper(unused: String) = doTest(KotlinUnwrappers.KotlinCatchUnwrapper::class.java)
    fun doTestCatchRemover(unused: String) = doTest(KotlinUnwrappers.KotlinCatchRemover::class.java)
    fun doTestFinallyUnwrapper(unused: String) = doTest(KotlinUnwrappers.KotlinFinallyUnwrapper::class.java)
    fun doTestFinallyRemover(unused: String) = doTest(KotlinUnwrappers.KotlinFinallyRemover::class.java)
    fun doTestLambdaUnwrapper(unused: String) = doTest(KotlinLambdaUnwrapper::class.java)
    fun doTestFunctionParameterUnwrapper(unused: String) = doTest(KotlinFunctionParameterUnwrapper::class.java)

    private fun doTest(unwrapperClass: Class<out Unwrapper>) {
        val testFile = testDataFile()

        myFixture.configureByFile(testFile)

        val fileText = file.text
        val isApplicableString = InTextDirectivesUtils.findStringWithPrefixes(fileText, "// IS_APPLICABLE: ")
        val isApplicableExpected = isApplicableString == null || isApplicableString == "true"
        val option = InTextDirectivesUtils.findStringWithPrefixes(fileText, "// OPTION: ")
        val optionIndex = option?.toInt() ?: 0

        val unwrappersWithPsi = KotlinUnwrapDescriptor().collectUnwrappers(project, editor, file)
        if (isApplicableExpected) {
            val selectedUnwrapperWithPsi = unwrappersWithPsi[optionIndex]
            assertEquals(unwrapperClass, selectedUnwrapperWithPsi.second.javaClass)
            val first = selectedUnwrapperWithPsi.first

            executeCommand(project, "Test") {
                runWriteAction {
                    selectedUnwrapperWithPsi.second.unwrap(editor, first)
                }
            }

            val expectedFile = File(testFile.parent, testFile.name + ".after")
            myFixture.checkResultByFile(expectedFile)
        } else {
            assert(unwrappersWithPsi.all { it.second.javaClass != unwrapperClass })
        }
    }
}